exporter.go

```diff
@@ -63,75 +66,133 @@ func hashNameAndLabels(name string, labels prometheus.Labels) uint64 {
+	strBuf.WriteString(name)
+	for _, labelName := range labelsNames(labels) {
+		strBuf.WriteByte(255)
+		strBuf.WriteString(labelName)
+	}
+	hash.Write(strBuf.Bytes())
+	return hash.Sum64()
+}
+
 type CounterContainer struct {
-	Elements map[uint64]prometheus.Counter
+	//               metric     labels names hash
+	Elements     map[string]map[uint64]*prometheus.CounterVec
 }
 
 func NewCounterContainer() *CounterContainer {
 	return &CounterContainer{
-		Elements: make(map[uint64]prometheus.Counter),
+		Elements:     make(map[string]map[uint64]*prometheus.CounterVec),
 	}
 }
 
 func (c *CounterContainer) Get(metricName string, labels prometheus.Labels, help string) (prometheus.Counter, error) {
-	hash := hashNameAndLabels(metricName, labels)
-	counter, ok := c.Elements[hash]
+	hash := hashNameAndLabelNames(metricName, labels)
+	var counter *prometheus.CounterVec
+	_, ok := c.Elements[metricName]
 	if !ok {
-		counter = prometheus.NewCounter(prometheus.CounterOpts{
-			Name:        metricName,
-			Help:        help,
-			ConstLabels: labels,
-		})
+		counter = prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: metricName,
+			Help: help,
+		}, labelsNames(labels))
 		if err := prometheus.Register(counter); err != nil {
 			return nil, err
 		}
-		c.Elements[hash] = counter
+		c.Elements[metricName] = map[uint64]*prometheus.CounterVec{
+			hash: counter,
+		}
+	}
+	counter, ok = c.Elements[metricName][hash]
+	if !ok {
+		err := fmt.Errorf(regErrNames, labelsNames(labels))
+		return nil, err
+	}
+
+	return counter.With(labels), nil
+}
+
+func (c *CounterContainer) Delete(metricName string, labels prometheus.Labels) {
+	if _, ok := c.Elements[metricName]; ok {
+		hash := hashNameAndLabelNames(metricName, labels)
+		c.Elements[metricName][hash].Delete(labels)
 	}
-	return counter, nil
 }
 
 type GaugeContainer struct {
-	Elements map[uint64]prometheus.Gauge
+	Elements map[string]map[uint64]*prometheus.GaugeVec
 }
 
 func NewGaugeContainer() *GaugeContainer {
 	return &GaugeContainer{
-		Elements: make(map[uint64]prometheus.Gauge),
+		Elements: make(map[string]map[uint64]*prometheus.GaugeVec),
 	}
 }
 
 func (c *GaugeContainer) Get(metricName string, labels prometheus.Labels, help string) (prometheus.Gauge, error) {
-	hash := hashNameAndLabels(metricName, labels)
-	gauge, ok := c.Elements[hash]
+	hash := hashNameAndLabelNames(metricName, labels)
+	var gauge *prometheus.GaugeVec
+	_, ok := c.Elements[metricName]
 	if !ok {
-		gauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		gauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
 			Name:        metricName,
 			Help:        help,
-			ConstLabels: labels,
-		})
+		}, labelsNames(labels))
 		if err := prometheus.Register(gauge); err != nil {
 			return nil, err
 		}
-		c.Elements[hash] = gauge
+		c.Elements[metricName] = map[uint64]*prometheus.GaugeVec{
+			hash: gauge,
+		}
+	}
+	gauge, ok = c.Elements[metricName][hash]
+	if !ok {
+		err := fmt.Errorf(regErrNames, labelsNames(labels))
```

Comment:
should we handle the mismatch gracefully at this point? a label set can be "expanded" with empty values without changing semantics.

Reply:
It seems that the only purpose of hashNameAndLabelNames and the second dimension in Elements field is to catch this type of error =)

I got this, need some time to remove unnecessary hashing.

Background story: It would be great if golang prometheus client has lastUpdateTime field for each label values set in MetricVec and has methods to list and delete individual sets. That will reduce memory and cpu footprint. For now there is «write only» behavior and we force to create intermediate layer to store label values.
I've implemented this algorithm in our other product and there is ConstLabels option in use, that is where hash for label names comes from.

Reply:
Keeping the client library write-only was a deliberate design decision, so we'll have to deal with it either way. The accepted way to solve this without the memory duplication is to implement the Collector interface and send ephemeral ConstMetric objects instead of setting the value on persistent metrics. The graphite exporter [does this](https://github.com/prometheus/graphite_exporter/blob/75f0db17d0be85c83d16855a82acffe31e605179/main.go#L196-L200). I don't expect you to do that refactoring in this PR though, that can happen separately later. Your work actually moves us closer to it.

Reply:
OK